### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by aliraza4638**
New Pull request

### CODE REVIEW

File: app/assets/builds/tailwind.css [hunk 1]
```diff
@@ -0,0 +1 @@
...35/var(--tw-text-opacity))}.shadow-xl{--tw-shadow:0 20px 25px -5px rgba(0,0,0,.1),0 8px 10px -6px rgba(0,0,0,.1);--tw-shadow-colored:0 20px 25px -5px var(--tw-shadow-color),0 8px 10px -6px var(--tw-shadow-color);box-shadow:var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow)}.hover\:bg-blue-600:hover{--tw-bg-opacity:1;background-color:rgb(37 99 235/var(--tw-bg-opacity))}.hover\:bg-blue-500:hover{--tw-bg-opacity:1;background-color:rgb(59 130 246/var(--tw-bg-opacity))}.hover\:underline:hover{-webkit-text-decoration-line:underline;text-decoration-line:underline}.focus\:outline-none:focus{outline:2px solid transparent;outline-offset:2px}.focus\:ring:focus{--tw-ring-offset-shadow:var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);--tw-ring-shadow:var(--tw-ring-inset) 0 0 0 calc(3px + var(--tw-ring-offset-width)) var(--tw-ring-color);box-shadow:var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow,0 0 #0000)}.focus\:ring-blue-300:focus{--tw-ring-opacity:1;--tw-ring-color:rgb(147 197 253/var(--tw-ring-opacity))}.active\:bg-blue-500:active{--tw-bg-opacity:1;background-color:rgb(59 130 246/var(--tw-bg-opacity))}
```
Comment (HassanAkbar) [inline app/assets/builds/tailwind.css:1]: @aliraza4638 Is this a compiled file for local build ?
If yes, should it be in `.gitignore` ?

File: app/assets/stylesheets/application.tailwind.css [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+/*
+
+@layer components {
+  .btn-primary {
+    @apply py-2 px-4 bg-blue-200;
+  }
+}
+
+*/
```
Comment (HassanAkbar) [inline app/assets/stylesheets/application.tailwind.css]: >Commented code is there which is not required.

@aliraza4638 we shouldn't have any unused commented code as we discussed already ?

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    @post = Post.find(params[:post_id])
+    @comment = @post.comments.new(comment_params)
+
+    @comment.user_id = current_user.id
+
+    respond_to do |format|
+      if @comment.save
+
+        format.turbo_stream
+        # format.html { redirect_to post_comments_path(@post), notice: "Comment was successfully created." }
```
Comment (HassanAkbar) [inline app/controllers/comments_controller.rb]: Unused code. Please ensure removal of such code everywhere.

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        format.html { redirect_to post_path(params[:post_id]), notice: 'Comment was successfully created.' }
+        format.json { render :show, status: :created, location: @comment }
```
Comment (HassanAkbar) [inline app/controllers/comments_controller.rb]: Are you using json, html and turbo_stream together ?
Lets not add any unused functionality ?

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        format.html { render :new, status: :unprocessable_entity }
+        format.json { render json: @comment.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # PATCH/PUT /comments/1 or /comments/1.json
+  def update
+    respond_to do |format|
+      if @comment.update(comment_params)
```
Comment (HassanAkbar) [inline app/controllers/comments_controller.rb:55]: Feels like anyone can update anyone's comment. You need to handle authorization properly for all actions. Use Pundit for that.

File: app/controllers/comments_controller.rb [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      else
+        format.html { render :edit, status: :unprocessable_entity }
+        format.json { render json: @comment.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # DELETE /comments/1 or /comments/1.json
+  def destroy
+    @comment.destroy
```
Comment (HassanAkbar) [inline app/controllers/comments_controller.rb]: > I can't see error handling is not done in cases such as destroying of records. 

@aliraza4638 The comments are not resolved. If you won't resolve comments and ask for re-review that will create frustration for the reviewer. The clients/devs won't like this at all.

Please ensure all comments I sent earlier are taken care of at all places.

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  end
+
+  # GET /posts/new
+  def new
+    @post = Post.new
+  end
+
+  # GET /posts/1/edit
+  def edit
+    content_not_found if current_user.id != @post.user_id
```
Comment (HassanAkbar) [inline app/controllers/posts_controller.rb]: Handling this with edit and ignoring update and other actions will be a huge risk as stated above. Use Pundit.

File: app/views/comments/_comment.html.erb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+<%= turbo_frame_tag comment do %>
+  <div id="<%= dom_id comment %>">
+    <div class="shadow-xl mb-4 p-4">
+      <p>
+        <strong>Content:</strong>
+        <%= comment.content %>
+      </p>
+      <div class="flex flex-row">
+        <% if comment.user_id == current_user.id %>
```
Comment (HassanAkbar) [inline app/views/comments/_comment.html.erb]: @aliraza4638 Adding conditions like this is always risky. Ideally there should be a single method which should be used by Pundit for authorization ?

File: app/views/devise/shared/_links.html.erb [hunk 9]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
+  <%= link_to "Log in", new_session_path(resource_name), class: " p-0  text-white border-solid border-4 rounded-lg border-blue-400 bg-blue-400 hover:underline active:bg-blue-500 focus:outline-none focus:ring focus:ring-blue-300 mb-8", style:'width: fit-content;' %><br />
+<% end %>
+
+<%- if devise_mapping.registerable? && controller_name != 'registrations' %>
+  <%= link_to "Sign up", new_registration_path(resource_name), class: " p-0  text-white border-solid border-4 rounded-lg border-blue-400 bg-blue-400 hover:underline active:bg-blue-500 focus:outline-none focus:ring focus:ring-blue-300 mb-8", style:'width: fit-content;' %><br />
+<% end %>
+
+<%- if devise_mapping.recoverable? && controller_name != 'passwords' && controller_name != 'registrations' %>
+  <%= link_to "Forgot your password?", new_password_path(resource_name), class: " p-0  text-white border-solid border-4 rounded-lg border-blue-400 bg-blue-400 hover:underline active:bg-blue-500 focus:outline-none focus:ring focus:ring-blue-300 mb-8 ", style:'width: fit-content;' %><br />
```
Comment (HassanAkbar) [inline app/views/devise/shared/_links.html.erb]: 1. Avoid Inline styling.
2. Extra spacing at start of class

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -5,18 +5,39 @@
...
+    <div class="bg-blue-600 py-4  sticky top-0 grid grid-cols-2">
+      <div class="text-start ml-4 text-4xl font-mono 	tracking-wide ">
+        <p>StackOverFlow</p>
+      </div>
+      <div class="flex flex-row-reverse">
+        
+        <p class="text-white	">
+          <% if user_signed_in? %>
+            Logged in as <strong><%= current_user.email %></strong>.
+            <%= link_to 'Edit profile', edit_user_registration_path, :class => 'px-8 hover:underline bg-blue-600 border-solid border-4 rounded-lg border-blue-400' %>
```
Comment (HassanAkbar) [inline app/views/layouts/application.html.erb]: Avoid using old hash syntax

File: app/views/posts/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<div class="grid-rows" style="width: 1000px">
+  <h2 class="font-medium leading-tight text-4xl mt-0 mb-2 text-blue-600">New Post</h2>
+  <%= render "form", post: @post %>
+  <br>
+  <div>
+    <%= link_to "Back to posts", posts_path, class: "text-blue-600	 hover:underline " %>
```
Comment (HassanAkbar) [inline app/views/posts/new.html.erb]: spacing ?

File: db/migrate/20220913131537_add_user_to_post.rb [hunk 12]
```diff
@@ -0,0 +1,5 @@
+class AddUserToPost < ActiveRecord::Migration[7.0]
+  def change
+    add_reference :posts, :user, null: true, foreign_key: true
```
Comment (HassanAkbar) [inline db/migrate/20220913131537_add_user_to_post.rb]: Can a post be added without a user ?

File: db/migrate/20220913114218_create_comments.rb [hunk 13]
```diff
@@ -0,0 +1,10 @@
+class CreateComments < ActiveRecord::Migration[7.0]
+  def change
+    create_table :comments do |t|
+      t.references :post, null: false, foreign_key: true
```
Comment (HassanAkbar) [inline db/migrate/20220913114218_create_comments.rb:4]: @aliraza4638 Whats the significance of `foreign_key: true` ?

File: test/models/comment_test.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class CommentTest < ActiveSupport::TestCase
+  # test "the truth" do
```
Comment (HassanAkbar) [inline test/models/comment_test.rb]: Lets add test cases for models or remove un-needed files if you feel there is no need to test case for any specific model ?

File: app/views/devise/shared/_links.html.erb (hunk 9 above)
Comment (aliraza4638) [reply app/views/devise/shared/_links.html.erb]: Inline style was added because width class property was not working as expected.

Comment (HassanAkbar) [reply app/views/devise/shared/_links.html.erb]: Do read up on inline styling vs classes (Pros vs Cons) ? 
Everything with style can be achieved with classes as well.

File: app/controllers/comments_controller.rb [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    @comment.user_id = current_user.id
+    authorize @comment
+
+    respond_to do |format|
+      if @comment.save
+
+        format.turbo_stream
+        format.html { redirect_to post_path(params[:post_id]), notice: 'Comment was successfully created.' }
+      else
+        format.html { render :new, status: :unprocessable_entity }
```
Comment (HassanAkbar) [inline app/controllers/comments_controller.rb]: We should display proper error by using comment.errors.full_messages in this case as well ?

File: app/controllers/comments_controller.rb [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        format.html { render :edit, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # DELETE /comments/1 or /comments/1.json
+  def destroy
+    authorize @comment
+    respond_to do |format|
+      @comment.destroy!
```
Comment (HassanAkbar) [inline app/controllers/comments_controller.rb]: @aliraza4638 this is fine too but you can also use if & else just like above to follow consistency. Do read up `destroyed?`

File: app/controllers/comments_controller.rb [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+      format.html { redirect_to post_path(params[:post_id]), notice: 'Comment was successfully destroyed.' }
+    rescue StandardError
+      format.html { redirect_to post_path(params[:post_id]), alert: 'Comment was not destroyed.' }
```
Comment (HassanAkbar) [inline app/controllers/comments_controller.rb]: Rails convention is to use alert for warning and error for errors.

File: app/helpers/comments_helper.rb [hunk 18]
```diff
@@ -0,0 +1,2 @@
+module CommentsHelper
```
Comment (HassanAkbar) [inline app/helpers/comments_helper.rb]: Should we remove unnecessary helpers?

File: app/policies/comment_policy.rb [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+  def index?
+    true
+  end
+
+  def create?
+    user.present?
+  end
+
+  def update?
+    return true if user.present? && user == comment.user
```
Comment (HassanAkbar) [inline app/policies/comment_policy.rb:13]: Quick Challenge: How can we refactor this line in Ruby way ?

File: app/views/comments/_comment.html.erb [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+  <div id="<%= dom_id comment %>">
+    <div class="shadow-xl mb-4 p-4">
+      <p>
+        <strong>Content:</strong>
+        <%= comment.content %>
+      </p>
+      <div class="flex flex-row">
+        <% if policy(comment).update? %>
+          <%= link_to "Edit", edit_post_comment_path(comment.post_id, comment), class: " p-0 m-1 text-white border-solid border-4 rounded-lg border-blue-400 bg-blue-400 hover:underline active:bg-blue-500 focus:outline-none focus:ring focus:ring-blue-300 mb-8" %>
+          <%= button_to "Destroy", post_comment_path(comment.post_id, comment), method: :delete, form: { data: { turbo_confirm: 'Are you sure?' } }, class: " p-0 m-1 text-white border-solid border-4 rounded-lg border-blue-400 bg-blue-400 hover:underline active:bg-blue-500 focus:outline-none focus:ring focus:ring-blue-300 mb-8"%>
```
Comment (HassanAkbar) [inline app/views/comments/_comment.html.erb]: Let ensure spacing is fixed in all app.
```suggestion
          <%= link_to "Edit", edit_post_comment_path(comment.post_id, comment), class: "p-0 m-1 text-white border-solid border-4 rounded-lg border-blue-400 bg-blue-400 hover:underline active:bg-blue-500 focus:outline-none focus:ring focus:ring-blue-300 mb-8" %>
          <%= button_to "Destroy", post_comment_path(comment.post_id, comment), method: :delete, form: { data: { turbo_confirm: 'Are you sure?' } }, class: " p-0 m-1 text-white border-solid border-4 rounded-lg border-blue-400 bg-blue-400 hover:underline active:bg-blue-500 focus:outline-none focus:ring focus:ring-blue-300 mb-8" %>
```

File: app/views/comments/create.turbo_stream.erb [hunk 21]
```diff
@@ -0,0 +1 @@
+<%= turbo_stream.append "comments", @comment%>
```
Comment (HassanAkbar) [inline app/views/comments/create.turbo_stream.erb]: Lets fix spacing everywhere

**Pull request merged**